# **Colored Cubical Complexes, $C_{3,4}$** 
---

* This notebook imports $C_{3,4}$, the list of all 68 3-dimensional cubes colored with four colors.
* Given a target cube, C, and a collection of cubes, $\mathcal{C}$  (perhaps a subset of $C_{3, 4}$ or all of $C_{3, 4}$), containing eight or more distinct cubes that can contribute to a 2x2x2 build of C, we consider all subsets of 8 cubes from $\mathcal{C}$ and for each subset count the number of distinct solutions possible in a 2x2x2 build of C.


In [1]:
import csv

def csv_to_kvs(fileName):
    make_dictionary = {}
    with open(fileName,'r') as csvFile:
        csvreader = csv.reader(csvFile)
        for row in csvreader:
            #position 1 and following are integers
            make_dictionary[row[0]] = [int(val) for val in row[1:]]
        print(make_dictionary)
    return make_dictionary

C34=csv_to_kvs('C_34all.csv')  
print(C34)

print(len(C34.keys()))



{'111.243': [111, 113, 114, 112, 243, 134, 123, 142], '333.142': [333, 331, 332, 334, 243, 134, 123, 142], '444.123': [444, 441, 443, 442, 243, 134, 123, 142], '222.134': [222, 221, 223, 224, 243, 134, 123, 142], '221.433': [332, 334, 221, 223, 243, 134, 123, 142], '224.313': [331, 332, 223, 224, 243, 134, 123, 142], '113.422': [113, 112, 221, 224, 243, 134, 123, 142], '114.232': [114, 112, 221, 223, 243, 134, 123, 142], '114.233': [113, 114, 331, 332, 243, 134, 123, 142], '112.343': [113, 112, 331, 334, 243, 134, 123, 142], '223.144': [441, 442, 223, 224, 243, 134, 123, 142], '221.434': [443, 442, 221, 224, 243, 134, 123, 142], '331.244': [331, 334, 443, 442, 243, 134, 123, 142], '332.414': [332, 334, 441, 443, 243, 134, 123, 142], '112.344': [114, 112, 441, 443, 243, 134, 123, 142], '113.424': [113, 114, 441, 442, 243, 134, 123, 142], '113.244': [113, 114, 441, 442, 124, 132, 143, 234], '112.434': [114, 112, 441, 443, 124, 132, 143, 234], '332.144': [332, 334, 441, 443, 124, 132, 143

In [2]:
%run module2x2x2_ext_match.ipynb
from datetime import date

today = date.today()
print("Today's date:", today)

Today's date: 2022-09-18


We initially focus on the target 111.243 and look for the maximum solution number. The collection of cubes with at least 5 usable corners in the target cube are  111.243
113.422
114.232
114.233
112.343
112.344
113.424
113.214
112.314
112.413
113.423
114.324 and
112.432

In [11]:
SubdictionaryFor111243={key: C34[key] for key in ["111.243","113.422","114.232","114.233","112.343","112.344",
"113.424","113.423","114.324","112.432","113.214","112.314","112.413"]}   
#all cubes that have 5 or more usable corners to make the target 
print(len(SubdictionaryFor111243), SubdictionaryFor111243)
print(type(SubdictionaryFor111243))

13 {'111.243': [111, 113, 114, 112, 243, 134, 123, 142], '113.422': [113, 112, 221, 224, 243, 134, 123, 142], '114.232': [114, 112, 221, 223, 243, 134, 123, 142], '114.233': [113, 114, 331, 332, 243, 134, 123, 142], '112.343': [113, 112, 331, 334, 243, 134, 123, 142], '112.344': [114, 112, 441, 443, 243, 134, 123, 142], '113.424': [113, 114, 441, 442, 243, 134, 123, 142], '113.423': [113, 113, 132, 143, 234, 243, 134, 123], '114.324': [114, 114, 124, 143, 234, 243, 134, 142], '112.432': [112, 112, 124, 132, 234, 243, 123, 142], '113.214': [113, 114, 113, 114, 124, 132, 123, 142], '112.314': [114, 112, 114, 112, 132, 143, 134, 123], '112.413': [113, 112, 113, 112, 124, 143, 134, 142]}
<class 'dict'>


The SolutionFrequencyList function first calculates (and displays) the number of subsets of 8 cubes from the input subdictionary collection. Then for each subset of eight cubes, it counts the number of ways those eight cubes can be used to create a 2x2x2 version of the target cube. This 'number of ways' is the solution number for the target and the eight cubes. When this is count is complete for every collection of eight cubes, then the distribution of solution numbers is created 
'solution number':'number of collections of eight cubes that have that solution number'.

We see that there are 13!/(8!* 5!) = 1287 collections of 8 cubes and the cube 111.243 is the only cube that contains the corner 111, thus any collction of cubes without the target cube cannot be used to make the target.  Thus 12!/(8!* 4!) = 495 collections of 8 cubes will yield no solutions.


In [12]:
SolutionFrequencyList(SubdictionaryFor111243, C34["111.243"])  

There are 1287 collections to make the target [111, 113, 114, 112, 243, 134, 123, 142]
Checked 1287 collections of 8 cubes
495 of the collections yield no solutions
 1392 :  3
 1824 :  3
 1384 :  36
 1216 :  36
 1544 :  18
 1720 :  12
 1376 :  12
 1648 :  6
 1328 :  12
 1992 :  12
 1944 :  6
 1332 :  36
 1516 :  48
 1468 :  24
 1436 :  24
 928 :  18
 1684 :  24
 1548 :  24
 1604 :  24
 1660 :  12
 1716 :  24
 1960 :  12
 1400 :  60
 1636 :  24
 1304 :  3
 1408 :  3
 1512 :  6
 1432 :  18
 1752 :  6
 1608 :  6
 1624 :  6
 1448 :  3
 1936 :  3
 1364 :  12
 1316 :  12
 1228 :  12
 1496 :  12
 1464 :  12
 1504 :  12
 1480 :  12
 1256 :  30
 1612 :  12
 1488 :  12
 1368 :  12
 1668 :  12
 1564 :  12
 1320 :  24
 1288 :  12
 1344 :  3
 1416 :  6
 1312 :  3
 1248 :  6



The maximum solution numbers are  1992 :  12,
 1960 :  12,
 1944 :  6,
 1936 :  3,
 1824 :  3, and
 1752 :  6.  Let's find the collections that yeild the largest solution number.
 

In [20]:
LookMaxFor111243={key: C34[key] for key in ["111.243","113.422","114.232","114.233","112.343","112.344",
"113.424","113.214","112.314","112.413"]}   
#all cubes that have 6 or more usable corners to make the target 
print(len(LookMaxFor111243), LookMaxFor111243)
print(type(LookMaxFor111243))

10 {'111.243': [111, 113, 114, 112, 243, 134, 123, 142], '113.422': [113, 112, 221, 224, 243, 134, 123, 142], '114.232': [114, 112, 221, 223, 243, 134, 123, 142], '114.233': [113, 114, 331, 332, 243, 134, 123, 142], '112.343': [113, 112, 331, 334, 243, 134, 123, 142], '112.344': [114, 112, 441, 443, 243, 134, 123, 142], '113.424': [113, 114, 441, 442, 243, 134, 123, 142], '113.214': [113, 114, 113, 114, 124, 132, 123, 142], '112.314': [114, 112, 114, 112, 132, 143, 134, 123], '112.413': [113, 112, 113, 112, 124, 143, 134, 142]}
<class 'dict'>


In [21]:
SolutionFrequencyList(LookMaxFor111243, C34["111.243"])  

# Nine collections have no solutions because they do not contain the target cube, thus cannot fill the 111 corner.

There are 45 collections to make the target [111, 113, 114, 112, 243, 134, 123, 142]
Checked 45 collections of 8 cubes
9 of the collections yield no solutions
 1824 :  3
 1992 :  12
 1944 :  6
 1960 :  12
 1936 :  3


In [22]:
Search1MaxFor111243={key: C34[key] for key in ["111.243","113.422","114.232","114.233","112.344", "113.424",
"113.214", "112.413"]}   
#all cubes that have 6 or more usable corners to make the target 

test = Search1MaxFor111243
print(len(test), test)
print(type(test))

SolutionFrequencyList(Search1MaxFor111243, C34["111.243"])  

8 {'111.243': [111, 113, 114, 112, 243, 134, 123, 142], '113.422': [113, 112, 221, 224, 243, 134, 123, 142], '114.232': [114, 112, 221, 223, 243, 134, 123, 142], '114.233': [113, 114, 331, 332, 243, 134, 123, 142], '112.344': [114, 112, 441, 443, 243, 134, 123, 142], '113.424': [113, 114, 441, 442, 243, 134, 123, 142], '113.214': [113, 114, 113, 114, 124, 132, 123, 142], '112.413': [113, 112, 113, 112, 124, 143, 134, 142]}
<class 'dict'>
There are 1 collections to make the target [111, 113, 114, 112, 243, 134, 123, 142]
Checked 1 collections of 8 cubes
0 of the collections yield no solutions
 1992 :  1




Theorem:
* A max solution contains 111.243, and
* five of the six 113.422, 114.232, 114.233, 112.343, 112.344, 113.424, and 
* two of three 113.214, 112.314, 112.413, 
* WITH the condition that when the cube removed from the six contains corners 11a and 11b, then the cube from the three containing two copies of 11a and 11b cannot be removed.
* This shows why there are 12 = (6 choose 5)x2 ways to attain the maximum solution number 1992.



In [24]:
Search2MaxFor111243={key: C34[key] for key in ["111.243",
                                               "112.343","113.422","113.424","114.232","114.233",
"113.214", "112.314"]}   
#all cubes that have 6 or more usable corners to make the target 

test = Search1MaxFor111243
print(len(test), test)
print(type(test))

SolutionFrequencyList(Search2MaxFor111243, C34["111.243"]) 

8 {'111.243': [111, 113, 114, 112, 243, 134, 123, 142], '113.422': [113, 112, 221, 224, 243, 134, 123, 142], '114.232': [114, 112, 221, 223, 243, 134, 123, 142], '114.233': [113, 114, 331, 332, 243, 134, 123, 142], '112.344': [114, 112, 441, 443, 243, 134, 123, 142], '113.424': [113, 114, 441, 442, 243, 134, 123, 142], '113.214': [113, 114, 113, 114, 124, 132, 123, 142], '112.413': [113, 112, 113, 112, 124, 143, 134, 142]}
<class 'dict'>
There are 1 collections to make the target [111, 113, 114, 112, 243, 134, 123, 142]
Checked 1 collections of 8 cubes
0 of the collections yield no solutions
 1992 :  1


## Next search for minimum nonzero solution number.

All below is old.

In [26]:
SubDict111243={"111.243":[111, 113, 114, 112, 243, 134, 123, 142]}
search_key=["111.", "112.", "113.", "114."]

for partial_key in search_key:
    SubDict111243.update(dict(filter(lambda item: partial_key in item[0], C34.items())))
    
print(SubDict111243, len(SubDict111243))

{'111.243': [111, 113, 114, 112, 243, 134, 123, 142], '111.234': [111, 113, 114, 112, 124, 132, 143, 234], '112.343': [113, 112, 331, 334, 243, 134, 123, 142], '112.344': [114, 112, 441, 443, 243, 134, 123, 142], '112.434': [114, 112, 441, 443, 124, 132, 143, 234], '112.433': [113, 112, 331, 334, 124, 132, 143, 234], '112.432': [112, 112, 124, 132, 234, 243, 123, 142], '112.314': [114, 112, 114, 112, 132, 143, 134, 123], '112.413': [113, 112, 113, 112, 124, 143, 134, 142], '112.334': [114, 112, 332, 334, 132, 143, 134, 123], '112.443': [113, 112, 443, 442, 124, 143, 134, 142], '113.422': [113, 112, 221, 224, 243, 134, 123, 142], '113.424': [113, 114, 441, 442, 243, 134, 123, 142], '113.244': [113, 114, 441, 442, 124, 132, 143, 234], '113.242': [113, 112, 221, 224, 124, 132, 143, 234], '113.423': [113, 113, 132, 143, 234, 243, 134, 123], '113.214': [113, 114, 113, 114, 124, 132, 123, 142], '113.224': [113, 114, 223, 224, 124, 132, 123, 142], '114.232': [114, 112, 221, 223, 243, 134, 123

In [18]:
SolutionFrequencyList(SubDict111243, C34["111.243"])   #max soln number 3228:12

There are 490314 collections to make the target [111, 113, 114, 112, 243, 134, 123, 142]
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
Checked 490314 collections of 8 cubes
220936 of the collections yield no solutions
 196 :  1824
 160 :  2154
 264 :  1896
 36 :  3810
 212 :  996
 176 :  924
 140 :  876
 156 :  5328
 112 :  2424
 132 :  3192
 240 :  1578
 24 :  9207
 108 :  5652
 192 :  3474
 336 :  822
 48 :  3496
 608 :  348
 252 :  631
 412 :  468
 604 :  216
 632 :  288
 584 :  192
 388 :  204
 468 :  360
 60 :  3900
 384 :  780
 476 :  336
 516 :  384
 844 :  186
 768 :  132
 696 :  318
 500 :  96
 780 :  216
 616 :  342
 316 :  672
 460 :  144
 464 :  288
 304 :  702
 360 :  1704
 300 :  768
 580 :  156
 556 :  168
 520 :  168
 96 :  2256
 404 :  624
 852 :  120
 836 :  168
 800 :  246
 536 :  312
 732 :  129
 144 :  3324
 596 :  132
 728 :  204
 640 :  168
 432 :  924
 744 :  228
 120 :  6063
 480 :  216
 84 :  7218
 356 :  426

In [24]:
SubDict112334={"112.334":[114, 112, 332, 334, 132, 143, 134, 123]}
search_key=["112.", "114.", "332.", "334."]

for partial_key in search_key:
    SubDict112334.update(dict(filter(lambda item: partial_key in item[0], C34.items())))
    
print(SubDict112334, len(SubDict112334))

{'112.334': [114, 112, 332, 334, 132, 143, 134, 123], '112.343': [113, 112, 331, 334, 243, 134, 123, 142], '112.344': [114, 112, 441, 443, 243, 134, 123, 142], '112.434': [114, 112, 441, 443, 124, 132, 143, 234], '112.433': [113, 112, 331, 334, 124, 132, 143, 234], '112.432': [112, 112, 124, 132, 234, 243, 123, 142], '112.314': [114, 112, 114, 112, 132, 143, 134, 123], '112.413': [113, 112, 113, 112, 124, 143, 134, 142], '112.443': [113, 112, 443, 442, 124, 143, 134, 142], '114.232': [114, 112, 221, 223, 243, 134, 123, 142], '114.233': [113, 114, 331, 332, 243, 134, 123, 142], '114.323': [113, 114, 331, 332, 124, 132, 143, 234], '114.322': [114, 112, 221, 223, 124, 132, 143, 234], '114.324': [114, 114, 124, 143, 234, 243, 134, 142], '332.414': [332, 334, 441, 443, 243, 134, 123, 142], '332.144': [332, 334, 441, 443, 124, 132, 143, 234], '332.412': [332, 332, 124, 132, 234, 243, 123, 142], '332.134': [332, 334, 332, 334, 132, 143, 134, 123], '334.214': [334, 334, 124, 143, 234, 243, 134

In [25]:
SolutionFrequencyList(SubDict112334, C34["112.334"])   #max soln number 2464: 11   NOT ACTUAL MAX! 2528 maybe

There are 75582 collections to make the target [114, 112, 332, 334, 132, 143, 134, 123]
.
.
.
.
.
.
.
Checked 75582 collections of 8 cubes
1199 of the collections yield no solutions
 152 :  382
 132 :  412
 160 :  538
 472 :  655
 200 :  737
 504 :  548
 744 :  243
 1008 :  126
 264 :  631
 48 :  62
 76 :  300
 212 :  326
 96 :  446
 328 :  529
 424 :  533
 66 :  40
 182 :  46
 80 :  355
 284 :  296
 364 :  185
 228 :  350
 244 :  284
 88 :  101
 100 :  412
 260 :  328
 372 :  251
 360 :  676
 300 :  481
 376 :  640
 404 :  238
 736 :  174
 332 :  221
 496 :  510
 664 :  252
 168 :  521
 704 :  304
 808 :  211
 600 :  396
 776 :  184
 232 :  641
 380 :  142
 628 :  130
 180 :  350
 520 :  364
 192 :  616
 176 :  698
 568 :  211
 816 :  286
 1088 :  83
 272 :  640
 720 :  238
 752 :  339
 800 :  180
 872 :  223
 1552 :  36
 680 :  229
 584 :  294
 888 :  177
 1168 :  69
 280 :  720
 832 :  203
 968 :  162
 1616 :  34
 648 :  262
 1936 :  18
 1080 :  208
 1296 :  37
 50 :  84
 142 :  92


In [51]:
OtherKeys112334={"112.334", "112.314", "112.343", "112.344", "112.433", "112.434", "111.243", "111.234", "113.423", "114.232", "114.233", "114.322", "114.323", "123.413", "221.431", "221.343","221.433","223.413", "331.421", "332.144", "332.414", "332.134", "332.412", "333.124", "333.142", "441.321", "443.213"}

print(len(OtherKeys112334))

OtherSubdict112334={key: C34[key] for key in OtherKeys112334}

SolutionFrequencyList(OtherSubdict112334, C34["112.334"])


27
There are 2220075 collections to make the target [114, 112, 332, 334, 132, 143, 134, 123]
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
.
Checked 2220075 collections of 8 cubes
163620 of the collections yield no solutions
 400 :  19353
 416 :  17034
 472 :  9378
 368 :  41712
 384 :  60366
 432 :  55925
 440 :  10428
 816 :  15908
 800 :  7659
 736 :  30066
 392 :  9846
 336 :  28512
 424 :  4644
 288 :  126681
 304 :  38070
 328 :  8226
 688 :  9936
 672 :  12436
 512 :  2934
 224 :  20526
 192 :  90955
 352 :  29982
 608 :  17103
 448 :  10149
 496 :  2988
 904 :  2970
 944 :  2310
 896 :  31

In [35]:
my_list=["112.334", "112.314", "112.343", "112.344", "112.433", "112.434", "111.243", "111.234", "113.423", "114.232", "114.233", "114.322", "114.323", "123.413", "221.431", "221.343","221.433","223.413", "331.421", "332.144", "332.414", "332.134", "332.412", "333.124", "333.142", "441.321", "443.213"]

print(len(my_list))

27


In [37]:
A_collection=["112.334","112.314","332.134","111.234", "111.243", "113.423", "221.431", "332.414", "333.124"]
print(len(A_collection))

A_Subdict112334={key: C34[key] for key in A_collection}
SolutionFrequencyList(A_Subdict112334, C34["112.334"])

9
There are 9 collections to make the target [114, 112, 332, 334, 132, 143, 134, 123]
Checked 9 collections of 8 cubes
0 of the collections yield no solutions
 1472 :  4
 1824 :  2
 736 :  2
 864 :  1


In [50]:
B_collection=["112.334", "112.314", "332.134","112.432","113.214", "331.234", "331.432", "332.412", "334.214"]
print(len(B_collection))
B_Subdict112334={key: C34[key] for key in B_collection}
SolutionFrequencyList(B_Subdict112334, C34["112.334"])

9
There are 9 collections to make the target [114, 112, 332, 334, 132, 143, 134, 123]
Checked 9 collections of 8 cubes
0 of the collections yield no solutions
 2080 :  2
 2528 :  2
 1472 :  2
 1280 :  2
 1152 :  1


In [ ]:
B_collection=["112.334", "112.314", "332.134", "112.432","113.214","114.324", "331.234", "331.432", "332.412","334.214"]
112.334
112.314
332.134
112.413 -->out
112.432
113.214
114.324-->out
331.234
331.432
332.412
334.214

In [5]:
#solution numbers for 112.343 in Ko
Ko=["111.243", "333.142", "444.123", "222.134", "221.433", "224.313", "113.422", "114.232", "114.233", "112.343", "223.144", "221.434", "331.244", "332.414", "112.344", "113.424"]
Subdict112343Ko={key: C34[key] for key in Ko}

SolutionFrequencyList(Subdict112343Ko, C34["112.343"])


There are 12870 collections to make the target [113, 112, 331, 334, 243, 134, 123, 142]
.
Checked 12870 collections of 8 cubes
886 of the collections yield no solutions
 288 :  448
 384 :  498
 624 :  196
 144 :  772
 336 :  240
 240 :  404
 264 :  496
 72 :  616
 168 :  680
 120 :  464
 216 :  248
 480 :  330
 192 :  674
 360 :  376
 504 :  440
 96 :  783
 840 :  124
 528 :  84
 816 :  232
 48 :  588
 600 :  244
 24 :  200
 648 :  128
 432 :  268
 696 :  88
 408 :  176
 1152 :  20
 1032 :  112
 672 :  96
 1368 :  40
 1296 :  44
 960 :  111
 1008 :  146
 576 :  146
 456 :  88
 792 :  96
 744 :  40
 312 :  272
 1344 :  100
 720 :  180
 1824 :  16
 1248 :  42
 768 :  46
 864 :  33
 936 :  64
 912 :  20
 1128 :  56
 1680 :  8
 2136 :  16
 1464 :  8
 1056 :  24
 2088 :  12
 1560 :  16
 1440 :  12
 1608 :  40
 1176 :  8
 2568 :  4
 2040 :  12
 1968 :  8
 1776 :  2
 1920 :  16
 1512 :  16
 1536 :  18
 984 :  16
 1200 :  18
 1320 :  8
 1272 :  12
 2832 :  4
 1848 :  8
 1392 :  4
 1872 :  6
 2

In [6]:
SolutionFrequencyList(Subdict112343Ko, C34["112.344"])

There are 12870 collections to make the target [114, 112, 441, 443, 243, 134, 123, 142]
.
Checked 12870 collections of 8 cubes
886 of the collections yield no solutions
 24 :  200
 48 :  588
 96 :  783
 192 :  674
 72 :  616
 144 :  772
 168 :  680
 120 :  464
 288 :  448
 264 :  496
 480 :  330
 384 :  498
 216 :  248
 336 :  240
 504 :  440
 360 :  376
 816 :  232
 528 :  84
 648 :  128
 240 :  404
 624 :  196
 432 :  268
 600 :  244
 840 :  124
 696 :  88
 1152 :  20
 408 :  176
 672 :  96
 576 :  146
 1296 :  44
 960 :  111
 720 :  180
 768 :  46
 912 :  20
 1032 :  112
 1344 :  100
 864 :  33
 1128 :  56
 1824 :  16
 312 :  272
 744 :  40
 792 :  96
 456 :  88
 1008 :  146
 1368 :  40
 936 :  64
 1248 :  42
 1680 :  8
 552 :  80
 888 :  20
 1056 :  24
 1776 :  2
 1320 :  8
 1176 :  8
 1608 :  40
 2088 :  12
 984 :  16
 1464 :  8
 1920 :  16
 2568 :  4
 1440 :  12
 1872 :  6
 1560 :  16
 1392 :  4
 1848 :  8
 2496 :  2
 2136 :  16
 2616 :  4
 2040 :  12
 2832 :  4
 1200 :  18
 1272